tests/pipeline/pipeline.cpp

```diff
@@ -0,0 +1,356 @@
+        OutputData outputData;
+
+        for( const auto& data: results )
+        {
+             const InputData& inputData = data;
+             outputData.thanksForAllTheFish += inputData.meaningOfLife + addMoreFish;
+        }
+
+        output.set( "TestOutputData", outputData );
+    }
+
+    livre::DataInfos getInputDataInfos() const final
+    {
+        return
+        {
+            { "TestInputData", livre::getType< InputData >( )}
+        };
+    }
+
+    livre::DataInfos getOutputDataInfos() const final
+    {
+        return
+        {
+            { "TestOutputData", livre::getType< OutputData >( )}
+        };
+    }
+};
+
+class ConvertFilter : public livre::Filter
+{
+    void execute( const livre::FutureMap& input, livre::PromiseMap& output ) const final
+    {
+        const std::vector< OutputData >& results = input.get< OutputData >( "ConvertInputData" );
+
+        InputData inputData;
+        for( const auto& data: results )
+        {
+             inputData.meaningOfLife = data.thanksForAllTheFish + addMoreFish;
+        }
+
+        output.set( "ConvertOutputData", inputData );
+    }
+
+    livre::DataInfos getInputDataInfos() const final
+    {
+        return
+        {
+            { "ConvertInputData", livre::getType< OutputData >( )}
+        };
+
+    }
+
+    livre::DataInfos getOutputDataInfos() const final
+    {
+        return
+        {
+            { "ConvertOutputData", livre::getType< InputData >( )}
+        };
+    }
+};
+
+bool check_error( const std::runtime_error& ) { return true; }
+
+BOOST_AUTO_TEST_CASE( testFilterNoInput )
+{
+    livre::PipeFilterT< TestFilter > pipeFilter( "Producer" );
+
+    // Execute will fail because there are no inputs where data is retrieved
+    BOOST_CHECK_EXCEPTION( pipeFilter.execute(), std::runtime_error, check_error );
+    const livre::UniqueFutureMap portFutures( pipeFilter.getPostconditions( ));
+
+    // Results of the filter will be empty.
+    BOOST_CHECK_EXCEPTION( portFutures.get< OutputData >( "TestOutputData" ),
+                           std::runtime_error, check_error );
+}
+
+BOOST_AUTO_TEST_CASE( testFilterWithInput )
+{
+    livre::PipeFilterT< TestFilter > pipeFilter( "Producer" );
+    const uint32_t inputValue = 90;
+
+    pipeFilter.getPromise( "TestInputData" ).set( InputData( inputValue ));
+    pipeFilter.execute();
+
+    const livre::UniqueFutureMap portFutures( pipeFilter.getPostconditions( ));
+    const OutputData& outputData =
+            portFutures.get< OutputData >( "TestOutputData" );
+    BOOST_CHECK_EQUAL( outputData.thanksForAllTheFish, 151 );
+}
+
+BOOST_AUTO_TEST_CASE( testSetAndGetWrongParameters )
+{
+    TestFilter filter;
+    livre::PipeFilterT< TestFilter > pipeFilter( "Producer" );
+    pipeFilter.getPromise( "TestInputData" ).set( InputData());
+
+    BOOST_CHECK_EXCEPTION( pipeFilter.getPromise( "InputData" ).set( OutputData( 0 )),
+                           std::runtime_error, check_error );
```

Comment:
I will keep this one as a runtime error. It is not the logic but it is really runtime incompatible types.